homeassistant/components/shelly/update.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+from typing import Any, Final
+
+from homeassistant.components.update import (
+    UpdateDeviceClass,
+    UpdateEntity,
+    UpdateEntityDescription,
+    UpdateEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import BLOCK, CONF_SLEEP_PERIOD, DATA_CONFIG_ENTRY, DOMAIN
+from .entity import (
+    RestEntityDescription,
+    RpcEntityDescription,
+    ShellyRestAttributeEntity,
+    ShellyRpcAttributeEntity,
+    async_setup_entry_rest,
+    async_setup_entry_rpc,
+)
+from .utils import get_device_entry_gen
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class RpcUpdateRequiredKeysMixin:
+    """Class for RPC update required keys."""
+
+    installed_version: Callable[[dict, dict | None], Any]
+    latest_version: Callable[[dict, dict | None], Any]
+    install: Callable
+
+
+@dataclass
+class RestUpdateRequiredKeysMixin:
+    """Class for REST update required keys."""
+
+    installed_version: Callable[[dict, dict | None], Any]
+    latest_version: Callable[[dict, dict | None], Any]
+    install: Callable
+
+
+@dataclass
+class RpcUpdateDescription(
+    RpcEntityDescription, UpdateEntityDescription, RpcUpdateRequiredKeysMixin
+):
+    """Class to describe a RPC update."""
+
+
+@dataclass
+class RestUpdateDescription(
+    RestEntityDescription, UpdateEntityDescription, RestUpdateRequiredKeysMixin
+):
+    """Class to describe a REST update."""
+
+
+REST_UPDATES: Final = {
+    "fwupdate": RestUpdateDescription(
+        name="Firmware Update",
+        key="fwupdate",
+        installed_version=lambda status, shelly: status["update"]["old_version"],
+        latest_version=lambda status, shelly: status["update"]["new_version"],
+        install=lambda wrapper: wrapper.async_trigger_ota_update(),
+        device_class=UpdateDeviceClass.FIRMWARE,
+        entity_category=EntityCategory.CONFIG,
+        entity_registry_enabled_default=True,
+    ),
+    "fwupdate_beta": RestUpdateDescription(
+        name="Beta Firmware Update",
+        key="fwupdate",
+        installed_version=lambda status, shelly: status["update"]["old_version"],
+        latest_version=lambda status, shelly: status["update"].get("beta_version"),
+        install=lambda wrapper: wrapper.async_trigger_ota_update(beta=True),
+        device_class=UpdateDeviceClass.FIRMWARE,
+        entity_category=EntityCategory.CONFIG,
+        entity_registry_enabled_default=False,
+    ),
+}
+
+RPC_UPDATES: Final = {
+    "fwupdate": RpcUpdateDescription(
+        name="Firmware Update",
+        key="sys",
+        sub_key="available_updates",
+        installed_version=lambda status, shelly: shelly["ver"]
```

Comment:
Since the installed version is identical for both entities I think we can remove it and hardcode it in the entity property

Reply:
done